instant-xml-macros/src/de.rs

```diff
@@ -0,0 +1,271 @@
+use proc_macro2::{Ident, Span, TokenStream};
+use quote::quote;
+
+use crate::{get_namespaces, retrieve_attr};
+
+struct Tokens<'a> {
+    enum_: &'a mut TokenStream,
+    consts_: &'a mut TokenStream,
+    names_: &'a mut TokenStream,
+    match_: &'a mut TokenStream,
+}
+
+pub struct Deserializer {
+    fn_vec: Vec<TokenStream>,
+}
+
+impl Deserializer {
+    pub fn new(input: &syn::DeriveInput) -> Deserializer {
+        let ident = &input.ident;
+        let name = ident.to_string();
+        let mut fn_vec = Vec::new();
+
+        let (_, other_namespaces) = get_namespaces(&input.attrs);
+        let mut namespaces_map: TokenStream = proc_macro::TokenStream::from(
+            quote!(let mut namespaces_map = std::collections::HashMap::new();),
+        )
+        .into();
+        for (k, v) in other_namespaces.iter() {
+            namespaces_map.extend(quote!(
+                namespaces_map.insert(#k, #v);
+            ))
+        }
+
+        // Elements
+        let mut elements_enum = TokenStream::new();
+        let mut elements_lets = TokenStream::new();
+        let mut elements_names = TokenStream::new();
+        let mut elem_type_match = TokenStream::new();
+        let mut elements_tokens = Tokens {
+            enum_: &mut elements_enum,
+            consts_: &mut elements_lets,
+            names_: &mut elements_names,
+            match_: &mut elem_type_match,
+        };
+
+        // Attributes
+        let mut attributes_enum = TokenStream::new();
+        let mut attributes_lets = TokenStream::new();
+        let mut attributes_names = TokenStream::new();
+        let mut attr_type_match = TokenStream::new();
+        let mut attributes_tokens = Tokens {
```

Comment:
Since `Tokens` does not seem to escape this scope, it's unclear to me why `Tokens` holds `&mut` references to its members instead of just taking ownership. I'm guessing you could probably `derive(Default)` on `Tokens` and then call `Tokens::default()` here, drastically reducing the amount of boilerplate here.

Reply:
It is more about passing the `Tokens` into the `quote!`, because `Tokens` do not implements the ToTokens, and I need separete variables so I will not change it into stream. As far as I understand, even if `Tokens` would keep the values, I would need to assign them to `let some_token :TokenStream`, it seems as boilerplaty as current solution.  I migth by missing something though.

Reply:
Yeah, I guess we need bindings to `Tokens`'s members in order to splice them into the token stream. However, would still help not to have to repeat all the `TokenStream::new()` initialization.

Reply:
done